ENH: Update ITK from v4.13 to v5.0rc1, Update BRAINSTools and SimpleITK

phcerdan opened the pull request:
To help extension developers transition their code base, a migration guide is also
available on the wiki.
See https://www.slicer.org/wiki/Documentation/Nightly/Developers/Tutorials/MigrationGuide/ITK4-to-ITK5

This commit implements the following changes

- Compatibility with refactor of threading models in ITK: Fix compilation
  using ProcessingThreader class instead of PlatformMultiThreader.
  ITKv5 reorganized the multi-threading capabilities of ITK into a more modular organization
  and naming to support several modern parallelism paradigms.

  Note that these changes allow Slicer to compile with ITKv5, but does not take advantage
  of any enhancement supporting new threading models. ProcessingThreader
  is still the old-renamed PlatformMultiThreader.

  These changes were originally associated with PR https://github.com/Slicer/Slicer/pull/1057

- Fix compilation issue related to ComputeJacobianWithRespectToPosition parameter mismatch.

  ITKv5 change the argument of some transform functions from JacobianType to JacobianPositionType.
  This fixes the new signatures.

  These changes were originally associated with PR https://github.com/Slicer/Slicer/pull/1055

List of BRAINSTools changes:

```
$ git shortlog --no-merges b045970..ac4dafaa

[...]
```

List of SimpleITK changes:

```
$ git shortlog --no-merges v1.1.0..dc5c1a1d
[...]
```

Co-authored-by: Pablo Hernandez-Cerdan <pablo.hernandez.cerdan@outlook.com>
Co-authored-by: Hans Johnson <hans.j.johnson@gmail.com>
Co-authored-by: Sam Horvath <sam.horvath@kitware.com>
Co-authored-by: Jean-Christophe Fillion-Robin <jcfr@kitware.com>

hjmjohnson commented:
@phcerdan Thanks for pushing this forward.  It is long overdue.

phcerdan commented on Modules/CLI/ExpertAutomatedRegistration/ExpertAutomatedRegistration.cxx:
```diff
@@ -698,7 +698,11 @@ int main( int argc, char * argv[] )
+#if ITK_VERSION_MAJOR >= 5
+    itk::MultiThreaderBase::SetGlobalDefaultNumberOfThreads(numberOfThreads);
+#else
+     itk::MultiThreader::SetGlobalDefaultNumberOfThreads(numberOfThreads);
```
Thanks, solved!

jcfr commented:
Thanks :+1: 

I suggest to move the update of CMake version into its own Pr that will integrate asap.

The `Next steps` part of the commit message should be moved into the description part of this PR.

Here is a suggested commit message:

```
ENH: Update ITK from v4.13 to v5.0rc1, Update BRAINSTools

To help extension developers transition their code base, a migration guide is also
available on the wiki.
See https://www.slicer.org/wiki/Documentation/Nightly/Developers/Tutorials/MigrationGuide/ITK4-to-ITK5

This commit implements the following changes

- Compatibility with refactor of threading models in ITK: Fix compilation
  using ProcessingThreader class instead of PlatformMultiThreader.
  ITKv5 reorganized the multi-threading capabilities of ITK into a more modular organization
  and naming to support several modern parallelism paradigms.
  
  Note that these changes allow Slicer to compile with ITKv5, but does not take advantage
  of any enhancement supporting new threading models. ProcessingThreader
  is still the old-renamed PlatformMultiThreader.

  These changes were originally associated with PR https://github.com/Slicer/Slicer/pull/1057

- Fix compilation issue related to ComputeJacobianWithRespectToPosition parameter mismatch.

  ITKv5 change the argument of some transform functions from JacobianType to JacobianPositionType.
  This fixes the new signatures.

  These changes were originally associated with PR https://github.com/Slicer/Slicer/pull/1055

List of BRAINSTools changes:

[...]


Co-authored-by: Hans Johnson <hans.j.johnson@gmail.com>
```

jcfr commented:
Should SimpleITK be also updated at the same time ?

dzenanz commented:
I think that SimpleITK build against both ITKv4 and ITKv5. If so, it does not have to be updated. @blowekamp is that correct?

phcerdan commented:
CMake update of internal projects moved new PR: #1097

phcerdan commented:
Hey @dzenanz, maybe this is an issue for ITK, I am facing this:

With options:
```
    -DITK_LEGACY_REMOVE:BOOL=OFF   #<-- Allow LEGACY ITKv4 features for now.
    -DITK_LEGACY_SILENT:BOOL=OFF    #<-- DO NOT Silence ITKv5 migrations.
    -DModule_ITKDeprecated:BOOL=ON #<-- Needed for ITKv5 now. (itkMultiThreader.h and MutexLock backwards compatibility.)
```
```
CMake Error at Modules/Compatibility/V3Compatibility/itk-module-init.cmake:2 (message):
   Module_ITKV3Compatibility is ON, but ITKV3_COMPATIBILITY is OFF. Either turn Module_ITKV3Compatibility OFF or ITKV3_COMPATIBILITY ON
Call Stack (most recent call first):
  CMake/ITKModuleEnablement.cmake:318 (include)
  CMakeLists.txt:374 (include)

```

They are still there adding the following options:
```
    -DITKV3_COMPATIBILITY:BOOL=OFF
    -DModule_ITKV3Compatibility:BOOL=OFF
    -DITKV4_COMPATIBILITY:BOOL=ON
```

The only way to solve it is to remove the Module_ITKDeprecated

```
    -DITK_LEGACY_REMOVE:BOOL=OFF   #<-- Allow LEGACY ITKv4 features for now.
    -DITK_LEGACY_SILENT:BOOL=OFF  #<--- Do not silence
    -DModule_ITKDeprecated:BOOL=OFF
    -DITKV3_COMPATIBILITY:BOOL=OFF
    -DModule_ITKV3Compatibility:BOOL=OFF
    -DITKV4_COMPATIBILITY:BOOL=ON
```

blowekamp commented:
SimpleITK v1.1 will not work against ITKv5. It will need to be updated to v1.2.0 release. ( Note: there is also [this patch ] (https://github.com/SimpleITK/SimpleITK/commit/88f9e7bc45e97c6e47543c8d06de130b876f1f6a)on the SimpleITK release brach which addresses some build issues with the latest XCode which should be included too. )

phcerdan commented:
Thanks @blowekamp! I am testing with SimpleITK OFF for now. But will use v1.2.0 release plus the patch. By the way, your link contained a spurious: ~~)on~~ . Fixed: https://github.com/SimpleITK/SimpleITK/commit/88f9e7bc45e97c6e47543c8d06de130b876f1f6a

jcfr commented:
In addition of the [suggested](https://github.com/Slicer/Slicer/pull/1096#issuecomment-470164587) commit message update, consider adding @sjh26  as a co-author too:

```
Co-authored-by: Sam Horvath <sam.horvath@kitware.com>
```

phcerdan commented:
Hey @dzenanz, @jcfr are the two commits on top of BRAINSTools here: https://github.com/Slicer/BRAINSTools/commits/slicer-2017-12-09-v4.7.1-b045970 planned to be merged upstream?

jcfr commented:
> are the two commits on top of BRAINSTools here: https://github.com/Slicer/BRAINSTools/commits/slicer-2017-12-09-v4.7.1-b045970 planned to be merged upstream?

They have already. See BRAINSia/BRAINSTools@a9045b9b0c63315702cb63186b9c52cc6d2afbca and BRAINSia/BRAINSTools@a1e7fab7f2ec288fd9bb2f565e764bd030e61fa5

Next time, we will make sure to label the commit in the fork with `[backport]`

dzenanz commented:
Since you will be moving to the latest ITK, and don't need to provide option to build against ITK 4.x, you might as well move to the latest version of BRAINSTools. @hjmjohnson what do you think?

jcfr commented:
> you might as well move to the latest version of BRAINSTools. @hjmjohnson what do you think?

That is the plan. @phcerdan will push an updated branch shortly :rocket:

phcerdan commented:
Created branches in Slicer for [ITK](https://github.com/Slicer/ITK/tree/slicer-v5.0.rc01-2019-01-30-eb54c22), [BRAINSTools](https://github.com/Slicer/BRAINSTools/tree/slicer-v5.0.0-2019-01-20-ac4dafaa) and [SimpleITK](https://github.com/Slicer/SimpleITK/tree/slicer-v1.2.0-2019-02-27-dc5c1a1)

ITK: updated to rc01 + 1DCMTK patch
BRAINSTools: updated to current master
SimpleITK: update to 1.2.0 + patches (points to current SimpleITK release branch)

hjmjohnson commented:
FYI:  I'll do my best to make any changes to BRAINSTools that are necessary.

jcfr commented:
The rational behind creating branches in the Slicer forks is to avoid full rebuild when applying changes that are being reviewed in the upstream project.

Indeed, changing GIT_REPOSITORY causes the source tree to be removed and cloned again, whereas updating GIT_TAG lead to an update.

phcerdan commented:
Testing locally:
Windows:
Problems with ITK module: ITKDeprecated.
```cpp
"D:\T\Pablo\S-R\ITK-build\Modules\Compatibility\Deprecated\src\ITKDeprecated.vcxproj" (default target) (32) ->
         (Link target) -> 
    10>itkConditionVariable.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) const itk::ConditionVariable::`vftable'" (__imp_??_7ConditionVariable@itk@@6B@) referenced in function "protected: __cdecl itk::ConditionVariable::ConditionVariable(void)" (??0ConditionVariable@itk@@IEAA@XZ) [D:\T\Pablo\S-R\ITK-build\Modules\Compatibility\Deprecated\src\ITKDeprecated.vcxproj] [D:\T\Pablo\S-R\ITK.vcxproj]
    10>itkMutexLock.obj : error LNK2001: unresolved external symbol "public: virtual char const * __cdecl itk::SimpleMutexLock::GetNameOfClass(void)" (?GetNameOfClass@SimpleMutexLock@itk@@UEAAPEBDXZ) [D:\T\Pablo\S-R\ITK-build\Modules\Compatibility\Deprecated\src\ITKDeprecated.vcxproj] [D:\T\Pablo\S-R\ITK.vcxproj]
    10>itkMutexLock.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) const itk::SimpleMutexLock::`vftable'" (__imp_??_7SimpleMutexLock@itk@@6B@) referenced in function "public: __cdecl itk::SimpleMutexLock::SimpleMutexLock(void)" (??0SimpleMutexLock@itk@@QEAA@XZ) [D:\T\Pablo\S-R\ITK-build\Modules\Compatibility\Deprecated\src\ITKDeprecated.vcxproj] [D:\T\Pablo\S-R\ITK.vcxproj]
    10>D:\T\Pablo\S-R\ITK-build\bin\Release\ITKDeprecated-5.0.dll : fatal error LNK1120: 3 unresolved externals [D:\T\Pablo\S-R\ITK-build\Modules\Compatibility\Deprecated\src\ITKDeprecated.vcxproj] [D:\T\Pablo\S-R\ITK.vcxproj]
```
MacOS:
BRAINSTools requires an increase on the `OS_DEPLOYMENT_TARGET` to `10.11`:
`CMAKE_OS_DEPLOYMENT_TARGET=10.11`
(Rebuilding)

jcfr commented:
> MacOS: BRAINSTools requires an increase on the OS_DEPLOYMENT_TARGET to 10.11

For reference:

![image](https://user-images.githubusercontent.com/219043/53924739-41ab9700-404b-11e9-98fd-dbd8dc34bdca.png)

hjmjohnson commented:
> Note that in TBB 2018 and higher we do not support OS X 10.9, and it is not tested anymore. TBB team recommends you to either switch to 10.11 or stay at an older version of TBB.

As I am trying to incorporate newer versions of TBB to accelerate some processing, I choose to use 10.11 as the base deployment target.

hjmjohnson commented:
From the TBB Changelog file:

> Increased minimally supported version of macOS*
>    (MACOSX_DEPLOYMENT_TARGET) to 10.11.

jcfr commented:
> From the TBB Changelog file:

> >    Increased minimally supported version of macOS*
> >    (MACOSX_DEPLOYMENT_TARGET) to 10.11.

Makes sense. We will most likely target 10.11 as well.

To get an idea of the user base, I also setup a poll. When you have a chance, would appreciate your input:
https://discourse.slicer.org/t/which-version-of-macos-do-you-use-to-run-slicer/6055

jcfr commented:
Closing. This is superseded by https://github.com/Slicer/Slicer/pull/1103
